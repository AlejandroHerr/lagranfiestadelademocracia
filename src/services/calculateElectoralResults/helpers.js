import { parentRegion } from '../../constants/regions';
import calculateDistortion from '../../utils/calculateDistortion';
import compose from '../../utils/compose';
import get from '../../utils/get';
import updateImmutableArray from '../../utils/updateImmutableArray';

const sumAbsoluteData = (source, addition) => ({
  valids: source.valids + addition.valids,
  prevWhites: source.prevWhites + addition.prevWhites,
  prevNulls: source.prevNulls + addition.prevNulls,
  prevAbstentions: source.prevAbstentions + addition.prevAbstentions,
  prevVotes: source.prevVotes + addition.prevVotes,
  prevCensus: source.prevCensus + addition.prevCensus,
  whites: source.whites + addition.whites,
  nulls: source.nulls + addition.nulls,
  abstentions: source.abstentions + addition.abstentions,
  votes: source.votes + addition.votes,
  census: source.census + addition.census,
  seats: source.seats + addition.seats,
});

const addRelativeData = data => ({
  ...data,
  percVotes: data.votes / data.census,
  percAbstention: data.abstentions / data.census,
  percValid: data.valids / data.votes,
  percWhite: data.whites / data.votes,
  percNull: data.nulls / data.votes,
});

const sumElectoralData = compose(
  addRelativeData,
  sumAbsoluteData,
);

const sumResultsByParty = (source, addition, data) => {
  return source.concat(addition).reduce((total, party) => {
    const previousPartyIndex = total.findIndex(({ id }) => id === party.id);

    if (previousPartyIndex === -1) {
      const percVotes = party.votes / data.valids;
      const percSeats = party.seats / data.seats;
      const distortion = calculateDistortion(percVotes, percSeats);

      return total.concat({
        ...party,
        percVotes,
        percSeats,
        distortion,
      });
    }

    return updateImmutableArray(total, previousPartyIndex, previousParty => {
      const votes = previousParty.votes + party.votes;
      const percVotes = votes / data.valids;
      const seats = previousParty.seats + party.seats;
      const percSeats = seats / data.seats;
      const distortion = calculateDistortion(percVotes, percSeats);

      return {
        ...previousParty,
        votes,
        percVotes,
        seats,
        percSeats,
        distortion,
      };
    });
  }, []);
};

export const aggregateResults = (source, addition) => {
  const electoralData = sumElectoralData(source.electoralData, addition.electoralData);
  const resultsByParty = sumResultsByParty(source.resultsByParty, addition.resultsByParty, electoralData);

  return {
    ...source,
    electoralData,
    resultsByParty,
  };
};

export const sortBySeatsAndVotes = (partyA, partyB) =>
  partyB.seats === partyA.seats ? partyB.votes - partyA.votes : partyB.seats - partyA.seats;

export const getParentRegionType = results => {
  const regionType = get(results, [0, 'regionData', 'type']);
  return parentRegion[regionType];
};
