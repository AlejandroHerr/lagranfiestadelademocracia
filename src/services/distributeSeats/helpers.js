import calculateDistortion from '../../utils/calculateDistortion';

export const addPercVotesAndSeats = (results, { valids }) =>
  results.map(result => ({
    ...result,
    percVotes: result.votes / valids,
    seats: 0,
  }));

export const filterPartiesOverTreshold = (results, minThreshold) =>
  results.filter(({ percVotes }) => percVotes * 100 >= minThreshold);

export const addDistortion = (result, totalSeats) => {
  const percSeats = result.seats / totalSeats;

  return {
    ...result,
    percSeats,
    distortion: calculateDistortion(result.percVotes, percSeats),
  };
};
