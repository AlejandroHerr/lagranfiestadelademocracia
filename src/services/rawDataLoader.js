import mem from 'mem';

import rawParties from '../data/parties.json';
import rawRegions from '../data/regions.json';
import rawResults from '../data/results.json';
import matchFn from '../utils/matchFn';

const getParties = () => rawParties;

const getRegions = () => rawRegions;

const getRegionsByType = type => rawRegions.filter(matchFn(type, 'type'));

const getResults = () => rawResults;

const getResultsByType = type => {
  const regionForType = getRegionsByType(type);
  const regionIdsForType = regionForType.map(({ id }) => id);

  return rawResults
    .filter(({ regionId }) => regionIdsForType.includes(regionId))
    .map(result => {
      const regionData = regionForType.find(({ id }) => id === result.regionId);

      return {
        ...result,
        regionData,
      };
    });
};

export default {
  getParties,
  getRegions,
  getRegionsByType: mem(getRegionsByType),
  getResults,
  getResultsByType: mem(getResultsByType),
};
