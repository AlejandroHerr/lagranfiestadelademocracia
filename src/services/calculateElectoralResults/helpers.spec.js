import { aggregateResults } from './helpers';

import rawDataLoader from '../rawDataLoader';

const results = rawDataLoader.getResultsByType(0);

const prepareResult = result => ({
  ...result,
  resultsByParty: result.resultsByParty.map((resultForParty, idx) => ({
    ...resultForParty,
    seats: idx + 1,
  })),
});

describe('services/calculateErectoralResults/helpers', () => {
  it('should match exact snapshot', () => {
    expect(aggregateResults(prepareResult(results[0]), prepareResult(results[results.length - 1]))).toMatchSnapshot();
  });
});
