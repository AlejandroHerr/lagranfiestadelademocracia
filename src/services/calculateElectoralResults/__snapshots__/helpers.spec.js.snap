// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services/calculateErectoralResults/helpers should match exact snapshot 1`] = `
Object {
  "electoralData": Object {
    "abstentions": 20871910,
    "census": 69194076,
    "nulls": 451776,
    "percAbstention": 0.3016430192665627,
    "percNull": 0.009349249783215429,
    "percValid": 0.9906507502167846,
    "percVotes": 0.6983569807334373,
    "percWhite": 0.007388783027648223,
    "prevAbstentions": 22146632,
    "prevCensus": 73023696,
    "prevNulls": 454438,
    "prevVotes": 50877064,
    "prevWhites": 376264,
    "seats": 700,
    "valids": 47870390,
    "votes": 48322166,
    "whites": 357042,
  },
  "id": 99999,
  "regionData": Object {
    "alias": "total estatal",
    "id": 99999,
    "name": "Total Estatal",
    "oldId": 99999,
    "type": 0,
  },
  "regionId": 99999,
  "resultsByParty": Array [
    Object {
      "distortion": -0.3274591549628666,
      "id": 68,
      "percSeats": 0.002857142857142857,
      "percVotes": 0.33031629782000943,
      "seats": 2,
      "votes": 15812370,
    },
    Object {
      "distortion": -0.22092723527603836,
      "id": 77,
      "percSeats": 0.005714285714285714,
      "percVotes": 0.22664152099032409,
      "seats": 4,
      "votes": 10849418,
    },
    Object {
      "distortion": -0.2024038319184066,
      "id": 59,
      "percSeats": 0.008571428571428572,
      "percVotes": 0.21097526048983517,
      "seats": 6,
      "votes": 10099468,
    },
    Object {
      "distortion": -0.11908087292732372,
      "id": 13,
      "percSeats": 0.011428571428571429,
      "percVotes": 0.13050944435589515,
      "seats": 8,
      "votes": 6247538,
    },
    Object {
      "distortion": -0.012005861780409263,
      "id": 25,
      "percSeats": 0.014285714285714285,
      "percVotes": 0.026291576066123548,
      "seats": 10,
      "votes": 1258588,
    },
    Object {
      "distortion": -0.002988125704786255,
      "id": 10,
      "percSeats": 0.017142857142857144,
      "percVotes": 0.0201309828476434,
      "seats": 12,
      "votes": 963678,
    },
    Object {
      "distortion": 0.00804208614134959,
      "id": 16,
      "percSeats": 0.02,
      "percVotes": 0.01195791385865041,
      "seats": 14,
      "votes": 572430,
    },
    Object {
      "distortion": 0.015165874831125104,
      "id": 23,
      "percSeats": 0.022857142857142857,
      "percVotes": 0.007691268026017753,
      "seats": 16,
      "votes": 368184,
    },
    Object {
      "distortion": 0.02245214391849086,
      "id": 7,
      "percSeats": 0.025714285714285714,
      "percVotes": 0.0032621417957948537,
      "seats": 18,
      "votes": 156160,
    },
    Object {
      "distortion": 0.01667062726189255,
      "id": 41,
      "percSeats": 0.02857142857142857,
      "percVotes": 0.01190080130953602,
      "seats": 20,
      "votes": 569696,
    },
    Object {
      "distortion": 0.029266817576137808,
      "id": 81,
      "percSeats": 0.03142857142857143,
      "percVotes": 0.0021617538524336235,
      "seats": 22,
      "votes": 103484,
    },
    Object {
      "distortion": 0.03218495847403195,
      "id": 93,
      "percSeats": 0.03428571428571429,
      "percVotes": 0.0021007558116823365,
      "seats": 24,
      "votes": 100564,
    },
    Object {
      "distortion": 0.035188371290538,
      "id": 94,
      "percSeats": 0.037142857142857144,
      "percVotes": 0.0019544858523191477,
      "seats": 26,
      "votes": 93562,
    },
    Object {
      "distortion": 0.03812401779053816,
      "id": 5,
      "percSeats": 0.04,
      "percVotes": 0.0018759822094618407,
      "seats": 28,
      "votes": 89804,
    },
    Object {
      "distortion": 0.0417477723255888,
      "id": 49,
      "percSeats": 0.04285714285714286,
      "percVotes": 0.0011093705315540568,
      "seats": 30,
      "votes": 53106,
    },
    Object {
      "distortion": 0.04511729872504247,
      "id": 30,
      "percSeats": 0.045714285714285714,
      "percVotes": 0.0005969869892432462,
      "seats": 32,
      "votes": 28578,
    },
    Object {
      "distortion": 0.04806907210430975,
      "id": 17,
      "percSeats": 0.04857142857142857,
      "percVotes": 0.0005023564671188182,
      "seats": 34,
      "votes": 24048,
    },
    Object {
      "distortion": 0.051016542197140476,
      "id": 28,
      "percSeats": 0.05142857142857143,
      "percVotes": 0.00041202923143095346,
      "seats": 36,
      "votes": 19724,
    },
    Object {
      "distortion": 0.05397600300072162,
      "id": 85,
      "percSeats": 0.054285714285714284,
      "percVotes": 0.00030971128499266456,
      "seats": 38,
      "votes": 14826,
    },
    Object {
      "distortion": 0.05686661122131775,
      "id": 86,
      "percSeats": 0.05714285714285714,
      "percVotes": 0.00027624592153939,
      "seats": 40,
      "votes": 13224,
    },
    Object {
      "distortion": 0.059738293337489,
      "id": 8,
      "percSeats": 0.06,
      "percVotes": 0.00026170666251100103,
      "seats": 42,
      "votes": 12528,
    },
    Object {
      "distortion": 0.06271977192701257,
      "id": 52,
      "percSeats": 0.06285714285714286,
      "percVotes": 0.0001373709301302956,
      "seats": 44,
      "votes": 6576,
    },
    Object {
      "distortion": 0.06557971400931319,
      "id": 84,
      "percSeats": 0.06571428571428571,
      "percVotes": 0.00013457170497253103,
      "seats": 46,
      "votes": 6442,
    },
    Object {
      "distortion": 0.06844178684509211,
      "id": 53,
      "percSeats": 0.06857142857142857,
      "percVotes": 0.0001296417263364681,
      "seats": 48,
      "votes": 6206,
    },
    Object {
      "distortion": 0.07130381790139105,
      "id": 11,
      "percSeats": 0.07142857142857142,
      "percVotes": 0.00012475352718037183,
      "seats": 50,
      "votes": 5972,
    },
    Object {
      "distortion": 0.07417424663316331,
      "id": 9,
      "percSeats": 0.07428571428571429,
      "percVotes": 0.00011146765255098193,
      "seats": 52,
      "votes": 5336,
    },
    Object {
      "distortion": 0.07704647188257412,
      "id": 91,
      "percSeats": 0.07714285714285714,
      "percVotes": 0.00009638526028302673,
      "seats": 54,
      "votes": 4614,
    },
    Object {
      "distortion": 0.07992429558230046,
      "id": 44,
      "percSeats": 0.08,
      "percVotes": 0.00007570441769954245,
      "seats": 56,
      "votes": 3624,
    },
    Object {
      "distortion": 0.0827863266385994,
      "id": 2,
      "percSeats": 0.08285714285714285,
      "percVotes": 0.00007081621854344617,
      "seats": 58,
      "votes": 3390,
    },
    Object {
      "distortion": 0.08566481881000522,
      "id": 36,
      "percSeats": 0.08571428571428572,
      "percVotes": 0.00004946690428049573,
      "seats": 60,
      "votes": 2368,
    },
    Object {
      "distortion": 0.08853574889553706,
      "id": 35,
      "percSeats": 0.08857142857142856,
      "percVotes": 0.00003567967589150621,
      "seats": 62,
      "votes": 1708,
    },
    Object {
      "distortion": 0.09139318420903969,
      "id": 12,
      "percSeats": 0.09142857142857143,
      "percVotes": 0.000035387219531739763,
      "seats": 64,
      "votes": 1694,
    },
    Object {
      "distortion": 0.09425070308150225,
      "id": 51,
      "percSeats": 0.09428571428571429,
      "percVotes": 0.000035011204212040054,
      "seats": 66,
      "votes": 1676,
    },
    Object {
      "distortion": 0.09711269235832123,
      "id": 80,
      "percSeats": 0.09714285714285714,
      "percVotes": 0.000030164784535910403,
      "seats": 68,
      "votes": 1444,
    },
    Object {
      "distortion": 0.09997000233338396,
      "id": 39,
      "percSeats": 0.1,
      "percVotes": 0.000029997666616043864,
      "seats": 70,
      "votes": 1436,
    },
    Object {
      "distortion": 0.10282856569284568,
      "id": 88,
      "percSeats": 0.10285714285714286,
      "percVotes": 0.000028577164297178277,
      "seats": 72,
      "votes": 1368,
    },
    Object {
      "distortion": 0.10568754684710707,
      "id": 72,
      "percSeats": 0.10571428571428572,
      "percVotes": 0.000026738867178646342,
      "seats": 74,
      "votes": 1280,
    },
    Object {
      "distortion": 0.10854565063228916,
      "id": 38,
      "percSeats": 0.10857142857142857,
      "percVotes": 0.000025777939139413738,
      "seats": 76,
      "votes": 1234,
    },
    Object {
      "distortion": 0.11140479890447041,
      "id": 83,
      "percSeats": 0.11142857142857143,
      "percVotes": 0.00002377252410101526,
      "seats": 78,
      "votes": 1138,
    },
    Object {
      "distortion": 0.11426248489485283,
      "id": 34,
      "percSeats": 0.11428571428571428,
      "percVotes": 0.00002322939086144901,
      "seats": 80,
      "votes": 1112,
    },
    Object {
      "distortion": 0.11712819254538885,
      "id": 33,
      "percSeats": 0.11714285714285715,
      "percVotes": 0.000014664597468288852,
      "seats": 82,
      "votes": 702,
    },
    Object {
      "distortion": 0.11998587853577128,
      "id": 29,
      "percSeats": 0.12,
      "percVotes": 0.0000141214642287226,
      "seats": 84,
      "votes": 676,
    },
    Object {
      "distortion": 0.12284335562875387,
      "id": 79,
      "percSeats": 0.12285714285714286,
      "percVotes": 0.000013787228388989519,
      "seats": 86,
      "votes": 660,
    },
    Object {
      "distortion": 0.1257035483879343,
      "id": 24,
      "percSeats": 0.12571428571428572,
      "percVotes": 0.000010737326351425172,
      "seats": 88,
      "votes": 514,
    },
    Object {
      "distortion": 0.12856081658351703,
      "id": 27,
      "percSeats": 0.12857142857142856,
      "percVotes": 0.000010611987911525267,
      "seats": 90,
      "votes": 508,
    },
    Object {
      "distortion": 0.13141979773777843,
      "id": 1,
      "percSeats": 0.13142857142857142,
      "percVotes": 0.00000877369079299333,
      "seats": 92,
      "votes": 420,
    },
    Object {
      "distortion": 0.134282288368357,
      "id": 31,
      "percSeats": 0.13428571428571429,
      "percVotes": 0.0000034259173572640624,
      "seats": 94,
      "votes": 164,
    },
    Object {
      "distortion": 0.13714060105093895,
      "id": 87,
      "percSeats": 0.13714285714285715,
      "percVotes": 0.000002256091918198285,
      "seats": 96,
      "votes": 108,
    },
  ],
}
`;
